internal/database/badger/settings.go

```diff
@@ -0,0 +1,39 @@
+// Copyright 2022 ChainSafe Systems (ON)
+// SPDX-License-Identifier: LGPL-3.0-only
+
+package badger
+
+import (
+	"fmt"
+	"path/filepath"
+)
+
+// Settings is the database settings.
+type Settings struct {
+	// Path is the database directory path to use.
+	// It defaults to the current directory if left unset.
+	Path string
+	// InMemory is whether to use an in-memory database.
+	InMemory *bool
```

Comment:
Actually the point of this is to detect when it's left unset, since `false` is not 'unset'. That way we can set defaults as we want with a method.

More details on https://github.com/qdm12/gosettings#settings-struct I have been using this for about a year, and it has been quite scalable and nice to use.

Reply:
why do  we need to check if it is set or unset?

Reply:
To be able to set defaults if it's left unset (and it extends beyond eventually, such as merging/overriding settings structs)

Reply:
I don't see why this is necessary.  IMO `InMemory` should default to false.

Reply:
> To be able to set defaults if it's left unset (and it extends beyond eventually, such as merging/overriding settings structs).

I don't think I understand this completely.

>  it extends beyond eventually, such as merging/overriding settings structs

May be elaborate this part. Do we need to merge or override this setting at some point?

What I am really not able to understand is that why can't we treat false as unset. 

Say someone assigned a false value to inMemory, why do we care if it was set or it is there because it is default value?

Reply:
> What I am really not able to understand is that why can't we treat false as unset.

Because `false` or `true` are set values. The user can specify `false`.
Say we want to read from two configuration sources such as flags and toml, and we want flags to take precedence.
The flag specifies `--inmemory=false` and the toml config specifies `inmemory = true`. If we consider `false` as unset, then we would end up with a final value of `true` whereas it should be `false` due to precedence.
Same applies with merging/overriding settings.

It's perhaps out of scope and not used yet, but it's good code writing imo and there is no point neglecting this.

Reply:
> If we consider false as unset, then we would end up with a final value of true whereas it should be false due to precedence

If I want flag to take precedence, I would just take flag value and assign it to inMemory. It doesn't matter if it is set or not. So, if the flag says false, the value of inMemory will become false.

The way we deal with such setting is, 
- initialise with default settings with, say `Default()`.
- Override with say toml config (we don't need to know existing values in settings here)
- Override with say flag (we don't need existing values here as well)

I don't think we need inMemory to be pointer here.

Reply:
I don't see why we would want to support detecting an unset value for this attribute.  I see no reason why the default value for `InMemory` should not be `false`.

Reply:
Read https://github.com/ChainSafe/gossamer/pull/3005#discussion_r1071986475 again. Empty value should be invalid (and for a boolean, false and true are both 'set values').

> If I want flag to take precedence, I would just take flag value and assign it to inMemory. It doesn't matter if it is set or not. So, if the flag says false, the value of inMemory will become false.

Say the flag is `--inmemory=false`, then your field is set to false. Then your next source (i.e. toml) sets `inmemory = true`. How does it know if it should override this `false` with its `true`?? You can have ugly logic for every field to do this, but it's much more modular to just use a boolean pointer and if it's nil, then just set it to the value read from the settings source.